### PR TITLE
0-padding for kafka offset

### PR DESCRIPTION
**Pull request by willyborankin**

### CODE REVIEW

**Comment by willyborankin** [comment]
@ivanyu Could yo uplz take a look

File: src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java [hunk 1]
```diff
@@ -48,6 +48,16 @@
         FilenameTemplateVariable.START_OFFSET.name
     );
 
+    private static final Map<String, String> EXPECTED_VARIABLE_PARAMETERS =
```
Comment (ivanyu) [inline src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java]: This is not used, I guess can be removed

File: src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java [hunk 2]
```diff
@@ -96,11 +106,18 @@ public void put(final SinkRecord record) {
...
-            .bindVariable(FilenameTemplateVariable.PARTITION.name, () -> Integer.toString(tp.partition()))
-            .bindVariable(FilenameTemplateVariable.START_OFFSET.name, () -> Long.toString(headRecord.kafkaOffset()))
-            .render();
+            .bindVariable(
+                FilenameTemplateVariable.PARTITION.name,
+                () -> Integer.toString(tp.partition())
+            ).bindVariable(
+                FilenameTemplateVariable.START_OFFSET.name,
+                usePaddingParameter -> usePaddingParameter.asBoolean()
+                    ? String.format("%010d", headRecord.kafkaOffset())
```
Comment (ivanyu) [inline src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java]: Let's stick to 20, which is enough to accommodate even the largest offsets? [Here's how it's done in Kafka](https://github.com/apache/kafka/blob/b13a7ff7077f05b9dd21a3cd21fed6d0da5b8cdd/core/src/main/scala/kafka/log/Log.scala#L2477).

File: src/main/java/io/aiven/kafka/connect/gcs/templating/Template.java [hunk 3]
```diff
@@ -39,96 +41,111 @@
...
+                if (templatePart instanceof TextTemplatePart) {
+                    sb.append(((TextTemplatePart) templatePart).text());
+                } else if (templatePart instanceof VariableTemplatePart) {
+                    final VariableTemplatePart variableTemplatePart = (VariableTemplatePart) templatePart;
+                    final Function<VariableTemplatePart.Parameter, String> binding =
+                        bindings.get(variableTemplatePart.variableName());
                     // Substitute for bound variables, pass the variable pattern as is for non-bound.
-                    if (supplier != null) {
-                        sb.append(supplier.get());
+                    if (Objects.nonNull(binding)) {
```
[the remark this answers is not in the document]
Comment (willyborankin) [reply src/main/java/io/aiven/kafka/connect/gcs/templating/Template.java]: the reason is readability. for me readability looks better.

File: src/main/java/io/aiven/kafka/connect/gcs/config/FilenameTemplateValidator.java [hunk 4]
```diff
@@ -83,5 +90,28 @@ public final void ensureValid(final String name, final Object value) {
...
         }
+
+        boolean isVariableParametersSupported = true;
+        for (final Map.Entry<String, Map<String, Set<String>>> e
+            : SUPPORTED_VARIABLE_PARAMETERS_SET.entrySet()) {
+            if (template.variableParameters().containsKey(e.getKey())) {
+                final VariableTemplatePart.Parameter p =
+                    template.variableParameters().get(e.getKey());
+                if (e.getValue().containsKey(p.name())) {
+                    isVariableParametersSupported =
```
[the remark this answers is not in the document]
Comment (willyborankin) [reply src/main/java/io/aiven/kafka/connect/gcs/config/FilenameTemplateValidator.java]: this check comes after variables has been checked, so far I would prefer to keep them as is so far.

File: src/test/java/io/aiven/kafka/connect/gcs/templating/TemplateTest.java [hunk 5]
```diff
@@ -32,93 +31,116 @@
...
+    void variableFormatBothSpacesWithVariable() {
+        final Template te = Template.of("{{ foo:tt=true }}");
+        final Template.Instance instance = te.instance();
+        instance.bindVariable("foo", () -> "foo");
+        assertEquals("foo", instance.render());
+    }
+
+    @Test
+    void parseVariableWithParameter() {
+        Template.of("{{foo:tt=true}}")
```
[the remark this answers is not in the document]
Comment (willyborankin) [reply src/test/java/io/aiven/kafka/connect/gcs/templating/TemplateTest.java]: Should we throw exception in this case or it is part of GcsConfig file name pattern validation?

[the remark this answers is not in the document]
Comment (ivanyu) [reply src/test/java/io/aiven/kafka/connect/gcs/templating/TemplateTest.java]: I'd throw `IllegalArgumentException` from `Template.of` and wrap it in `ConfigException` in the validation.

File: src/main/java/io/aiven/kafka/connect/gcs/config/FilenameTemplateValidator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivanyu) [reply src/main/java/io/aiven/kafka/connect/gcs/config/FilenameTemplateValidator.java]: Gotcha

**Comment by willyborankin** [comment]
@ivanyu Could you please review 2 last commits

File: src/main/java/io/aiven/kafka/connect/gcs/templating/Template.java [hunk 6]
```diff
@@ -39,96 +41,111 @@
  * <p>Non-bound variables are left as is.
  */
 public final class Template {
```
[the remark this answers is not in the document]
Comment (ivanyu) [reply src/main/java/io/aiven/kafka/connect/gcs/templating/Template.java]: Still relevant :)

**Comment by ivanyu** [comment]
This is still relevant:
> Also, I'd ask you to add `padding=true` to the examples in README, like [here](https://github.com/aiven/aiven-kafka-connect-gcs#record-grouping) and [here](https://github.com/aiven/aiven-kafka-connect-gcs#configuration), to promote this as the recommended way.

File: README.md [hunk 7]
```diff
@@ -32,10 +32,14 @@ template for file names. It supports placeholders with variable names:
...
 - `topic` - the Kafka topic;
 - `partition` - the Kafka partition;
-- `start_offset` - the Kafka offset of the first record in the file;
+- `start_offset:padding=true|false` - the Kafka offset of the first record in the file, if `padding` sets to `true` will set leading zeroes for offset, default is `false`;
 - `key` - the Kafka key.
 
-Only the certain combinations of variables are allowed in the file name
+To 0-padding for kafka offset, you need to define additional parameter `padding` in the `start_offset` variable.
+The parameter accept boolean value `true` or `false`. For backward compatibility the default value is `false`.
+e.g. `{{topic}}-{{start_offset:padding=true}}.gz`, connector will create a file name: `topic-00000000000000000001.gz`
```
Comment (ivanyu) [inline README.md]: Let's put it this way:

To add zero padding to Kafka offsets, you need to add additional parameter `padding` in the `start_offset` variable, which value can be `true` or `false` (the default). For example: `{{topic}}-{{start_offset:padding=true}}.gz` will produce file names like `mytopic-00000000000000000001.gz`.

File: README.md [hunk 8]
```diff
@@ -35,9 +35,9 @@ template for file names. It supports placeholders with variable names:
 - `start_offset:padding=true|false` - the Kafka offset of the first record in the file, if `padding` sets to `true` will set leading zeroes for offset, default is `false`;
 - `key` - the Kafka key.
 
-To 0-padding for kafka offset, you need to define additional parameter `padding` in the `start_offset` variable.
-The parameter accept boolean value `true` or `false`. For backward compatibility the default value is `false`.
-e.g. `{{topic}}-{{start_offset:padding=true}}.gz`, connector will create a file name: `topic-00000000000000000001.gz`
+To add zero padding to Kafka offsets, you need to add additional parameter padding in the start_offset variable, 
+which value can be true or false (the default). 
+For example: {{topic}}-{{start_offset:padding=true}}.gz will produce file names like mytopic-00000000000000000001.gz.
```
Comment (ivanyu) [inline README.md]: Nice, but please add backticks where they belong :)

Comment (willyborankin) [reply README.md]: true. sorry

**Pull request merged**